File: solutions/c/rail-fence-cipher/1/rail_fence_cipher.c
#include "rail_fence_cipher.h"

#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char* encode(const char* plain, size_t rails) {
    size_t const len = strlen(plain);
    char* cipher = calloc(1, len + 1);
    size_t cipher_index = 0;
    for (size_t rail = 0; rail < rails; ++rail) {
        if (rail == 0 || rail == rails - 1) {
            for (size_t plain_index = rail; plain_index < len;
                 plain_index += 2 * (rails - 1)) {
                cipher[cipher_index] = plain[plain_index];
                ++cipher_index;
            }
        } else {
            size_t step = 2 * (rails - rail - 1);
            for (size_t plain_index = rail; plain_index < len;
                 plain_index += step, step = 2 * (rails - 1) - step) {
                cipher[cipher_index] = plain[plain_index];
                ++cipher_index;
            }
        }
    }

    return cipher;
}

char* decode(const char* cipher, size_t rails) {
    size_t const len = strlen(cipher);
    char* plain = calloc(1, len + 1);
    size_t cipher_index = 0;
    for (size_t rail = 0; rail < rails; ++rail) {
        if (rail == 0 || rail == rails - 1) {
            for (size_t plain_index = rail; plain_index < len;
                 plain_index += 2 * (rails - 1)) {
                plain[plain_index] = cipher[cipher_index];
                ++cipher_index;
            }
        } else {
            size_t step = 2 * (rails - rail - 1);
            for (size_t plain_index = rail; plain_index < len;
                 plain_index += step, step = 2 * (rails - 1) - step) {
                plain[plain_index] = cipher[cipher_index];
                ++cipher_index;
            }
        }
    }

    return plain;
}

File: solutions/haskell/rail-fence-cipher/1/src/RailFenceCipher.hs
{-# LANGUAGE BlockArguments #-}

module RailFenceCipher (encode, decode) where

import Control.Monad (when)
import Control.Monad.ST (ST, runST)
import Data.Foldable (for_, toList)
import Data.Function ((&))
import Data.Functor ((<&>))
import Data.STRef (STRef, modifySTRef, newSTRef, readSTRef)
import Data.Vector (Vector, freeze, (!))
import Data.Vector qualified as Vector
import Data.Vector.Mutable (MVector, new, write)

data Codec = Encode | Decode

encode :: Int -> String -> String
encode = subst Encode

decode :: Int -> String -> String
decode = subst Decode

subst :: Codec -> Int -> String -> String
subst codec rails messageStr =
    runST
        ( do
            out <- new $ length message
            cipherIx <- newSTRef 0
            for_ [0 .. rails - 1] \r -> do
                step <- newSTRef $ 2 * r
                plainIx <- newSTRef r
                while (plainIx <? length message) do
                    case codec of
                        Decode -> (out, plainIx) =. (message, cipherIx)
                        Encode -> (out, cipherIx) =. (message, plainIx)
                    modifySTRef cipherIx (+ 1)
                    whenM (step /=? maxStep) $ modifySTRef step (maxStep -)
                    plainIx += step
            freeze out
        )
        & toList
  where
    message = Vector.fromList messageStr
    maxStep = 2 * (rails - 1)

while :: (Monad m) => m Bool -> m () -> m ()
while cond body = go where go = whenM cond $ body *> go

whenM :: (Monad m) => m Bool -> m () -> m ()
whenM cond body = do c <- cond; when c body

(=.) :: (MVector s Char, STRef s Int) -> (Vector Char, STRef s Int) -> ST s ()
(dst, i) =. (src, j) = do
    i_ <- readSTRef i
    j_ <- readSTRef j
    write dst i_ $ src ! j_

(+=) :: STRef s Int -> STRef s Int -> ST s ()
x += y = do
    y_ <- readSTRef y
    modifySTRef x (+ y_)

(<?) :: (Ord a) => STRef s a -> a -> ST s Bool
x <? y = readSTRef x <&> (< y)

(/=?) :: (Eq a) => STRef s a -> a -> ST s Bool
x /=? y = readSTRef x <&> (/= y)
